crml/token-approvals/src/lib.rs

```diff
@@ -0,0 +1,153 @@
+pub trait Config: frame_system::Config {
+	/// The overarching event type.
+	type Event: From<Event<Self>> + Into<<Self as frame_system::Config>::Event>;
+	/// Handles a multi-currency fungible asset system
+	type MultiCurrency: MultiCurrency<AccountId = Self::AccountId, CurrencyId = AssetId, Balance = Balance>;
+	/// NFT ownership interface
+	type IsTokenOwner: IsTokenOwner<AccountId = Self::AccountId>;
+}
+
+impl<T: Config> OnTransferSubscriber for Module<T> {
+	/// Do anything that needs to be done after an NFT has been transferred
+	fn on_nft_transfer(token_id: &TokenId) {
+		// Set approval to none
+		Self::remove_erc721_approval(token_id);
+	}
+}
+
+decl_event!(
+	pub enum Event<T>
+	where
+		<T as frame_system::Config>::AccountId,
+		CollectionId = CollectionId,
+		SeriesId = SeriesId,
+		SerialNumber = SerialNumber,
+	{
+		// Approval has been set (account_id, (collection_id, series_id, serial_number))
+		NFTApprovalSet(Option<AccountId>, (CollectionId, SeriesId, SerialNumber)),
+		// Approval has been set for series (account_id, collection_id, series_id)
+		NFTApprovalSetForAll(AccountId, CollectionId, SeriesId),
+	}
+);
+
+decl_error! {
+	/// Error for the token-approvals module.
+	pub enum Error for Module<T: Config> {
+		/// The account is not the owner of the token
+		NotTokenOwner,
+		/// The caller account can't be the same as the operator account
+		CallerNotOperator,
+	}
+}
+
+// This module's storage items.
+decl_storage! {
+	trait Store for Module<T: Config> as TokenApprovals {
+		// Account with transfer approval for a single NFT
+		pub ERC721Approvals get(fn erc721_approvals): map hasher(twox_64_concat) (CollectionId, SeriesId, SerialNumber) => T::AccountId;
+		// Account with transfer approval for an NFT series of another account
+		pub ERC721ApprovalsForAll get(fn erc721_approvals_for_all): double_map hasher(twox_64_concat) T::AccountId, hasher(twox_64_concat) (CollectionId, SeriesId) => Vec<T::AccountId>;
+	}
+}
+
+// The module's dispatchable functions.
+decl_module! {
+	/// The module declaration.
+	pub struct Module<T: Config> for enum Call where origin: T::Origin {
+		// Initializing events
+		// this is needed only if you are using events in your module
+		fn deposit_event() = default;
+
+		/// Set approval for a single NFT
+		/// Mapping from token_id to operator
+		/// clears approval on transfer
+		#[weight = 125_000_000]
+		pub fn erc721_approval(
+			origin,
+			caller: T::AccountId,
+			operator_account: T::AccountId,
+			token_id: TokenId,
+		) -> DispatchResult {
+			// mapping(uint256 => address) private _tokenApprovals;
+
+			let _ = ensure_none(origin)?;
+			ensure!(caller != operator_account, Error::<T>::CallerNotOperator);
+			// Check that origin owns NFT
+			ensure!(T::IsTokenOwner::check_ownership(&caller, &token_id), Error::<T>::NotTokenOwner);
+			ERC721Approvals::<T>::insert(token_id, operator_account.clone());
+
+			Self::deposit_event(RawEvent::NFTApprovalSet(Some(operator_account), token_id));
```

Comment:
let's log the relevant event in the precompile so it's emmited for EVM based apps